src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1619,6 +1665,23 @@ private static bool CleanupReadStateAndCheckPending(State state, ReadState final
             return shouldComplete;
         }
 
+        internal async Task StartAsync(CancellationToken cancellationToken)
+        {
+            Debug.Assert(!Monitor.IsEntered(_state));
+
+            using CancellationTokenRegistration registration = cancellationToken.UnsafeRegister(static (s, token) =>
+            {
+                ((State)s!).StartCompletionSource.TrySetException(new OperationCanceledException(token));
+            }, _state);
+
+            // Fire of start of the stream
+            uint status = MsQuicApi.Api.StreamStartDelegate(_state.Handle, QUIC_STREAM_START_FLAGS.SHUTDOWN_ON_FAIL | QUIC_STREAM_START_FLAGS.INDICATE_PEER_ACCEPT);
+            QuicExceptionHelpers.ThrowIfFailed(status, "Could not start stream.");
+
+            // wait unit start completes.
+            await _state.StartCompletionSource.Task.ConfigureAwait(false);
```

Comment:
Now as we have it as a separate method, I think it could be just
```suggestion
            return new ValueTask(_state.StartCompletionSource.Task.WaitAsync(cancellationToken));
```
(and we won't need an additional CancellationTokenRegistration as well)

Reply:
I'm not trying to say what is right or wrong, just pointing out the consequences of such change.

Reply:
I think we can safely use `WaitAsync`, we should not get to coll `StartAsync` multiple times anyway since the stream instance gets immediately disposed if the task throws.